src/simsopt/solve/mpi.py

```diff
@@ -232,3 +235,276 @@ def _f_proc0(x):
+        # header of the output file until this first evaluation is
+        # done.
+        if not constraint_datalog_started:
+            # Initialize log file
+            constraint_datalog_started = True
+            datestr = datetime.now().strftime("%Y-%m-%d-%H-%M-%S")
+            constraint_file = open(f"constraint_{datestr}.dat", 'w')
+            constraint_file.write(f"Problem type:\nconstrained\nnparams:\n{prob.dof_size}\n")
+            constraint_file.write("function_evaluation,seconds")
+
+            for j in range(prob.dof_size):
+                constraint_file.write(f",x({j})")
+            constraint_file.write(",constraint_function\n")
+            for j in range(len(constraint_val)):
+                constraint_file.write(f",F({j})")
+            constraint_file.write("\n")
+
+        del_t = time() - start_time
+        constraint_file.write(f"{n_constraint_evals:6d},{del_t:12.4e}")
+        for xj in x:
+            constraint_file.write(f",{xj:24.16e}")
+        for fj in constraint_val:
+            constraint_file.write(f",{fj:24.16e}")
+        constraint_file.write("\n")
+        constraint_file.flush()
+
+        n_constraint_evals += 1
+        logger.debug(f"constraints are {constraint_val}")
+        return constraint_val
+
+    # prepare bounds 
+    if prob.has_bounds:
+        bounds = Bounds(prob.lb, prob.ub)
+    else:
+        bounds = None
+
+    # prepare linear constraints
+    constraints = []
+    if prob.has_lc:
+        lincon = LinearConstraint(prob.A_lc, lb=-np.inf, ub=prob.b_lc)
+        constraints.append(lincon)
+
+    # For MPI finite difference gradient, get the worker and leader action from
+    # MPIFiniteDifference
+    if grad:
+        with MPIFiniteDifference(prob.all_funcs, mpi, abs_step=abs_step,
+                                 rel_step=rel_step, diff_method=diff_method) as fd:
+
+            def obj_jac(x):
+                # dummy wrapper for batch finite difference
+                return fd.jac(x)[0]
+
+            if mpi.proc0_world:
+                if prob.has_nlc:
+                    def nlc_jac(x):
+                        # dummy wrapper for batch finite difference
+                        return fd.jac(x)[1:]
+                    #nlc = NonlinearConstraint(_nlc_proc0, lb=prob.lhs_nlc, ub=prob.rhs_nlc, jac=nlc_jac)
+                    nlc = NonlinearConstraint(_nlc_proc0, lb=-np.inf, ub=0.0, jac=nlc_jac)
+                    constraints.append(nlc)
+
+                # proc0_world does this block, running the optimization.
+                x0 = np.copy(prob.x)
+                logger.info("Using finite difference method implemented in "
+                            "SIMSOPT for evaluating gradient")
+                result = minimize(_f_proc0, x0, jac=obj_jac,
+                                  bounds=bounds, constraints=constraints,
+                                  method=opt_method, options=options)
+
```

Comment:
Ideally, 470 to 472 would be like
```
result = minimize(_f_proc0, x0, jac=obj_jac,
                  bounds=prob.bounds, constraints=prob.constraints,
                  method=opt_method, options=options)
```

Reply:
we can change it to 
`result = minimize(_f_proc0, x0, jac=obj_jac,
                  bounds=prob.bounds, constraints=constraints,
                  method=opt_method, options=options)`
but we cannot set `constraints = prob.constraints` because the nonlinear constraint handles must be generated inside of the solve function (the nonlinear constraints depend on the MPI finite difference jacobian).